### PR TITLE
Optional integration with c2isl

### PR DESCRIPTION
**Pull request by ezyang**
- We add c2isl as an optional dependency.  The recommended
  installation procedure is Conda; we'll pick up c2isl installs
  in $CONDA_PREFIX and use them if available.

- This commit set adds conversion functions between ATen tensors
  and dlpack.  The place they currently live is not long-term
  where we should put them, but as ATen does not presently have
  a dependency on the dlpack headers there isn't really anywhere
  we can put them at the moment.

- The code for dispatching to c2isl was heavily based off of
  the existing code for handling Caffe2.  We have a proof of
  concept operator using c2isl's MatMul implementation; gradient
  is NOT supported, and you must instantiate a new ISLMatMul class
  every time you want to run it on a different size.  You won't
  really get any performance increase this way, but this at least
  demonstrates it is possible to call to c2isl.

- In order to interface with c2isl, we have to allocate C-style
  DLTensor structures.  The memory management involved here is
  somewhat delicate and we have some custom deleters to handle
  the situation.

### CODE REVIEW

File: torch/csrc/autograd/c2isl_function.cpp [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+      break;
+    case at::ScalarType::Short:
+      t.code = DLDataTypeCode::kInt;
+      t.bits = 16;
+      break;
+    case at::ScalarType::Half:
+      t.code = DLDataTypeCode::kFloat;
+      t.bits = 16;
+      break;
+    case at::ScalarType::NumOptions:
```
Comment (apaszke) [inline torch/csrc/autograd/c2isl_function.cpp]: `default` seems to be a better choice for raising an error. Same below

File: torch/csrc/autograd/c2isl_function.cpp [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+      break;
+    default:
+      throw std::logic_error("Unsupported code " + std::to_string(t.code));
+  }
+  // TODO: This doesn't put the tensor on the correct device
+  return getType(backend, st);
+}
+
+at::Type& toATenType(const DLContext& ctx, const DLDataType& t) {
+  if (t.lanes != 1) throw std::logic_error("ATen does not support lanes != 1");
```
Comment (apaszke) [inline torch/csrc/autograd/c2isl_function.cpp]: It seems better to move this check to the overload above. This way it'd be called in both code paths + this one doesn't even use `t`

File: torch/csrc/autograd/c2isl_function.cpp [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+
+variable_list ISLFunction::apply(const variable_list& input_vars) {
+  AutoGIL gil;
+
+  // Use the same backend as the first tensor
+  auto& input = input_vars.at(0);
+  AutoGPU guard(input->data->getDevice());
+
+  if (!pImpl_) {
+    std::vector<const DLMetadata*> inMetas; // ownership managed by inTensorUPtrs
```
Comment (apaszke) [inline torch/csrc/autograd/c2isl_function.cpp]: There's no such thing as `inTensorUPtrs`  did you mean `inMetaUPtrs`?

File: torch/csrc/autograd/c2isl_function.cpp [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+    std::vector<DLMetadataUPtr> inMetaUPtrs;
+    for (auto input : input_vars) {
+      auto at = variableToATen(input);
+      inMetaUPtrs.emplace_back(toDLMetadata(at));
+      inMetas.emplace_back(inMetaUPtrs.back().get());
+    }
+    initTVM(inMetas);
+    pImpl_ = std::unique_ptr<c2isl::ISLTVMIROp>(
+              new c2isl::ISLTVMIROp(outputs, inputs, vars, ops)
```
Comment (apaszke) [inline torch/csrc/autograd/c2isl_function.cpp]: Can you append underscores to member names? I've just spent some time looking where are they defined. I know we haven't done that in most places in autograd, however `vars` seems more like a local variable name to me

File: torch/csrc/autograd/c2isl_function.cpp [hunk 5]
```diff
@@ -0,0 +1,321 @@
...
+  }
+
+  // OK, return outputs
+  return wrap_outputs(input_vars, std::move(compat_outputs), [&](FunctionFlags f) -> std::shared_ptr<Function> {
+    return std::make_shared<Error>("ISLFunction backwards not implemented yet", std::move(f));
+  });
+}
+
+void ISLMatMul::initTVM(const std::vector<const DLMetadata*>& input_metas) {
+  // TODO: I feel dirty doing copy-assignment
```
Comment (apaszke) [inline torch/csrc/autograd/c2isl_function.cpp]: this is move assigment

File: torch/csrc/autograd/c2isl_function.h [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+
+// DLTensor is a plain C object, so the onus is on us to correctly allocate
+// and free members.  An added complication is c2isl also uses the DLTensor
+// data type to represent size/stride information, WITHOUT an actual tensor.
+//
+// Here's our strategy:
+//  - We new-allocate shape and strides for both DLTensor and DLMetadata,
+//    and free them in our deleters.
+//  - The deleter for DLTensor also holds a reference to at::Tensor, to keep
```
Comment (apaszke) [inline torch/csrc/autograd/c2isl_function.h]: it holds an at::Tensor as a member, not as a reference

File: torch/csrc/autograd/functions/init.cpp [hunk 7]
```diff
@@ -262,6 +274,11 @@ bool THPAutograd_initFunctions(PyObject* _unused)
   static PyTypeObject CatClass;
   addClass<Cat, NoCtor>(module, CatClass, "Cat");
 
+  #ifdef WITH_C2ISL
+  static PyTypeObject ISLMatMulClass;
+  addClass<ISLMatMul, ISLMatMulCtor>(module, ISLMatMulClass, "ISLMatMul");
+  #endif
```
Comment (apaszke) [inline torch/csrc/autograd/functions/init.cpp]: This seems to be a debug code, and shouldn't be part of this PR

File: torch/lib/ATen/templates/TensorDerived.cpp [hunk 8]
```diff
@@ -26,13 +26,18 @@ IntList ${Tensor}::sizes() {
 int64_t ${Tensor}::dim() {
   if(isScalar())
     return 0;
-  return ${THTensor}_nDimension(${state,}tensor);
+  int64_t d = ${THTensor}_nDimension(${state,}tensor);
+  if (d != 0)
+    return d;
+  return kUndefinedDimensions;
```
Comment (apaszke) [inline torch/lib/ATen/templates/TensorDerived.cpp]: Why do we change the convention in here? All TH libs have 0-dim tensors

File: torch/csrc/autograd/c2isl_function.cpp [hunk 9]
```diff
@@ -0,0 +1,321 @@
...
+  vars = { ::tvm::Var("M"), ::tvm::Var("N"), ::tvm::Var("O") };
```
Comment (apaszke) [inline torch/csrc/autograd/c2isl_function.cpp]: What's O used for?

File: torch/csrc/autograd/c2isl_function.cpp (hunk 1 above)
Comment (ezyang) [inline torch/csrc/autograd/c2isl_function.cpp]: ATen's ScalarType is an enum class, so if we add default we lose exhaustiveness checking. I'll add a comment to this effect.

File: torch/csrc/autograd/c2isl_function.cpp (hunk 4 above)
Comment (ezyang) [inline torch/csrc/autograd/c2isl_function.cpp]: I gave them these names so we lined up with c2isl code better, but now that we have a Python interface this is less pressing, so I'll do this.

File: torch/csrc/autograd/c2isl_function.cpp (hunk 5 above)
Comment (ezyang) [inline torch/csrc/autograd/c2isl_function.cpp]: This code died.

File: torch/csrc/autograd/c2isl_function.cpp (hunk 9 above)
Comment (ezyang) [inline torch/csrc/autograd/c2isl_function.cpp]: This code died.

File: torch/lib/ATen/templates/TensorDerived.cpp (hunk 8 above)
Comment (ezyang) [inline torch/lib/ATen/templates/TensorDerived.cpp]: CC @killeent from who I got this patch.

File: torch/_isl.py [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+    @staticmethod
+    def name():
+        raise NotImplementedError
+
+
+# This is an example use-case of this functionality.
+class IslMatMul(IslFunction):
+    @staticmethod
+    def name():
+        return "matmul"
```
Comment (apaszke) [inline torch/_isl.py]: Is there any reason why we want to make it a method instead of a class attribute (`name = "matmul"`), or just use the class name?

Comment (ezyang) [inline torch/_isl.py]: Not really, I just picked something arbitrary. But one post facto justification is you might want to generate a name based on some parameters that are passed to the constructor; then you need a method.

Comment (apaszke) [inline torch/_isl.py]: Yes, except it can't be a static method that it is at the moment
